lib/tree.rb

```diff
@@ -1,64 +1,162 @@
 class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
-
-   def initialize(key, val)
+  
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
   def initialize
     @root = nil
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  # Time Complexity: Average O(log n), worst-case O(n)
+  # Space Complexity: 1
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def add_helper(curr_node, key, value)
+    if curr_node.nil?
+      return TreeNode.new(key, value) 
+    end 
+    if key <= curr_node.key
+      curr_node.left = add_helper(curr_node.left, key, value)
+    else 
+      curr_node.right = add_helper(curr_node.right, key, value)
+    end
+    return curr_node
+  end
+  
+  # Time Complexity: Average O(log n), worst-case O(n)
+  # Space Complexity: 1
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def find_helper(curr_node, key)
+    return if curr_node.nil?
+    if key == curr_node.key
+      return curr_node.value
+    elsif key < curr_node.key
+      find_helper(curr_node.left, key)
+    else 
+      find_helper(curr_node.right, key)
+    end
+  end 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(curr_node, list)
+    return list if curr_node.nil?
+    inorder_helper(curr_node.left, list)
+    list << {key: curr_node.key, value: curr_node.value}
+    inorder_helper(curr_node.right, list)
+    return list
+  end 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(curr_node, list)
+    return list if curr_node.nil?
+    list << {key: curr_node.key, value: curr_node.value}
+    preorder_helper(curr_node.left, list)
+    preorder_helper(curr_node.right, list)
+  end 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(curr_node, list)
+    return list if curr_node.nil?
+    postorder_helper(curr_node.left, list)
+    postorder_helper(curr_node.right, list)
+    list << {key: curr_node.key, value: curr_node.value}
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)...?
   def height
```

Comment:
Very compact and this works, but your space complexity is off.  It's O(log n) if the tree is balanced and O(n) if it's not.